src/ore/src/tracing/optimizer.rs

```diff
@@ -0,0 +1,313 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Tracing utilities for explainable plans.
+
+use std::fmt::Debug;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::path::PathBuf;
+use std::sync::Mutex;
+
+use tracing::{span, subscriber};
+use tracing_subscriber::{field, layer};
+
+/// A tracing layer used to accumulate a sequence of explainable plans.
+#[allow(missing_debug_implementations)]
+pub struct PlanTrace<T> {
+    path: Mutex<PathBuf>,
```

Comment:
Note that the `path` and `times` stack accumulated in the `Layer` implementation for `PlanTrace<T>` can be shared by a chain of `PlanTrace<T>` instances wrapped in a layered subscriber, but at the moment this work is replicated.

I didn't find a clean way to do this kind of bookkeeping in the `Registry` base of the layer chain.

Reply:
is `PathBuf` the right type here? or is just convenient based on how you implemented it? `String` might make more sense, as we aren't working with unix file paths

Reply:
I didn't realize that `PathBuf` is meant to abstract over OS paths and has `os` stuff inside. I will change this to `Vec<String>`.